### PR TITLE
Refactor chains/endpoint to use Option pattern 

### LINKED ISSUE #737: Refactor chains/endpoint to use Option pattern
## Motivation

We can simplify using of our chains with
[Option pattern](https://github.com/tmrts/go-patterns/blob/master/idiom/functional-options.md)

See at an example: https://github.com/networkservicemesh/sdk/commit/9007c4530373d301993ecd8985a1c9abebe359b2

Moreover, this can super help us with minimization of manual changes of upstream repositories.

Entry point: https://github.com/networkservicemesh/sdk/blob/master/pkg/networkservice/chains/endpoint/server.go

### PR DESCRIPTION
**Pull request by Tiberivs**
Refactor `chains/endpoint` to use Option pattern.
Closes #737.

### CODE REVIEW

File: pkg/networkservice/chains/endpoint/server.go [hunk 1]
```diff
@@ -54,28 +56,69 @@ type endpoint struct {
...
+		o.authzServer = authzServer
+	}
+}
+
+// WithTokenGenerator sets token generator function - generates tokens for use in Path
+func WithTokenGenerator(tokenGenerator token.GeneratorFunc) Option {
+	return func(o *serverOptions) {
+		o.tokenGenerator = tokenGenerator
+	}
+}
```
Comment (denis-tingaikin) [inline pkg/networkservice/chains/endpoint/server.go:88]: This is the required parameter for the endpoint...

File: pkg/networkservice/chains/endpoint/server.go [hunk 2]
```diff
@@ -54,28 +56,69 @@ type endpoint struct {
...
 // NewServer - returns a NetworkServiceMesh client as a chain of the standard Client pieces plus whatever
-//             additional functionality is specified
-//             - name - name of the NetworkServiceServer
-//             - authzServer authorization server chain element
-//             - tokenGenerator - token.GeneratorFunc - generates tokens for use in Path
-//             - additionalFunctionality - any additional NetworkServiceServer chain elements to be included in the chain
-func NewServer(ctx context.Context, name string, authzServer networkservice.NetworkServiceServer, tokenGenerator token.GeneratorFunc, additionalFunctionality ...networkservice.NetworkServiceServer) Endpoint {
+func NewServer(ctx context.Context, options ...Option) Endpoint {
+	opts := &serverOptions{
+		name:        "server-" + uuid.New().String(),
```
Comment (denis-tingaikin) [inline pkg/networkservice/chains/endpoint/server.go:100]:
```suggestion
		name:        "endpoint-" + uuid.New().String(),
```

File: pkg/networkservice/chains/nsmgrproxy/server.go [hunk 3]
```diff
@@ -22,30 +22,79 @@ import (
...
+// WithGRPCOptions sets gRPC Dial Options for the server
+func WithGRPCOptions(options ...grpc.DialOption) Option {
+	return func(o *serverOptions) {
+		o.grpcOptions = options
+	}
+}
+
 // NewServer creates new proxy NSMgr
-func NewServer(ctx context.Context, name string, authorizeServer networkservice.NetworkServiceServer, tokenGenerator token.GeneratorFunc, options ...grpc.DialOption) endpoint.Endpoint {
+func NewServer(ctx context.Context, options ...Option) endpoint.Endpoint {
```
Comment (denis-tingaikin) [inline pkg/networkservice/chains/nsmgrproxy/server.go:77]: Why this is changed in this PR?

**Review by denis-tingaikin — CHANGES_REQUESTED** [review]

File: pkg/networkservice/chains/nsmgrproxy/server.go [hunk 4]
```diff
@@ -22,30 +22,79 @@ import (
...
+
+// WithTokenGenerator sets token generator function for the server
+func WithTokenGenerator(function token.GeneratorFunc) Option {
+	return func(o *serverOptions) {
+		o.tokenGenerator = function
+	}
+}
+
+// WithGRPCOptions sets gRPC Dial Options for the server
+func WithGRPCOptions(options ...grpc.DialOption) Option {
```
Comment (denis-tingaikin) [inline pkg/networkservice/chains/nsmgrproxy/server.go:70]:
```suggestion
func WithDialOption(options ...grpc.DialOption) Option {
```

File: pkg/tools/sandbox/node.go [hunk 5]
```diff
@@ -95,10 +95,10 @@ func (n *Node) newEndpoint(
 	// 1. Create endpoint server
 	ep := endpoint.NewServer(
 		ctx,
-		nse.Name,
-		authorize.NewServer(authorize.Any()),
-		generatorFunc,
-		additionalFunctionality...,
+		endpoint.WithName(nse.Name),
+		endpoint.WithAuthorizationServer(authorize.NewServer(authorize.Any())),
```
Comment (denis-tingaikin) [inline pkg/tools/sandbox/node.go:99]:
```suggestion
```

**Review by denis-tingaikin — CHANGES_REQUESTED** [review]

File: pkg/networkservice/chains/endpoint/server.go (hunk 1 above)
Comment (Tiberivs) [reply pkg/networkservice/chains/endpoint/server.go]: removed from options

File: pkg/networkservice/chains/endpoint/server.go (hunk 2 above)
Comment (Tiberivs) [reply pkg/networkservice/chains/endpoint/server.go]: renamed

File: pkg/networkservice/chains/nsmgrproxy/server.go (hunk 4 above)
Comment (Tiberivs) [reply pkg/networkservice/chains/nsmgrproxy/server.go]: renamed

File: pkg/tools/sandbox/node.go (hunk 5 above)
Comment (Tiberivs) [reply pkg/tools/sandbox/node.go]: moved to default values

File: pkg/networkservice/chains/nsmgrproxy/server.go (hunk 3 above)
Comment (Tiberivs) [reply pkg/networkservice/chains/nsmgrproxy/server.go]: changes removed from this PR

**Review by denis-tingaikin — APPROVED** [review]

File: pkg/networkservice/chains/endpoint/server.go [hunk 6]
```diff
@@ -54,19 +56,52 @@ type endpoint struct {
...
+func WithName(name string) Option {
+	return func(o *serverOptions) {
+		o.name = name
+	}
+}
+
+// WithAuthorizeServer sets authorization server chain element
+func WithAuthorizeServer(authzServer networkservice.NetworkServiceServer) Option {
+	return func(o *serverOptions) {
+		o.authzServer = authzServer
```
Comment (denis-tingaikin) [inline pkg/networkservice/chains/endpoint/server.go:78]: We should prevent passing nil here.

**Review by denis-tingaikin — CHANGES_REQUESTED** [review]

File: pkg/networkservice/chains/endpoint/server.go (hunk 6 above)
Comment (Tiberivs) [reply pkg/networkservice/chains/endpoint/server.go]: Checking was added

**Review by denis-tingaikin — APPROVED** [review]

**Comment by denis-tingaikin** [comment]
@edwarnicke I think this PR is ready to merge. Do you have any comments on these changes?

**Pull request merged**